src/lib/player_cache.ts

```diff
@@ -15,186 +15,383 @@
+        let compatibility: any = to_old_style_player(new_style_player);
+        Object.assign(new_style_player, compatibility);
+
+        // If the data we're fed might be inconsistent with the current state
+        // of the player in question, then the caller will set dont_overwrite
+        // to be true. In this case, just return the new_style_player, as it
+        // will contain all the information the caller needs.
+        if (dont_overwrite) {
+            if (player_cache_debug) {
+                console.log("Converted old-style player without caching the result", player, new_style_player);
             }
-            delete active_fetches[player_id];
-            update(player);
-            if (required_fields) {
-                for (let field of required_fields) {
-                    if (!(field in cache[player.id])) {
-                        console.warn("Required field ", field, " was not resolved by fetch");
-                        cache[player.id][field] = "[ERROR]";
-                    }
+            return new_style_player;
+        }
+
+        // Copy any new or changed information to the cache. Note that this will
+        // update everybody's copy of the cached data. The data is cached both by
+        // username and by id. Along the way, we take note of whether any
+        // information is changed in the cache.
+        let changed = false;
+        let was_cached = true;
+        if (player_id !== 0) {
+            for (let name in new_style_player) {
+                if (cached[name] !== new_style_player[name]) {
+                    changed = true;
                 }
+                cached[name] = new_style_player[name];
             }
-            resolve(cache[player.id]);
-        })
-        .catch((err) => {
-            delete active_fetches[player_id];
-            console.error(err);
-            reject(err);
-        });
-    });
-}
 
+            was_cached = new_style_player.id in cache_by_id;
+            cache_by_id[new_style_player.id] = cached;
 
-export const player_cache = {
-    update: update,
-    lookup: lookup,
-    lookup_by_username: lookup_by_username,
-    nicknames: nicknames,
-    fetch: fetch,
-    watch: watch,
-    Listener: Listener,
-};
+            if (cached.username) {
+                if (!(cached.username in cache_by_username)) {
+                    nicknames.push(cached.username);
+                }
+                cache_by_username[cached.username] = cached;
+            }
+        }
 
-export default player_cache;
+        // If the cached information has changed, then inform everyone who is
+        // subscribed to the player.
+        if (changed && !dont_overwrite) {
+            publisher.publish(cached.id.toString(), cached);
+        }
+
+        // If the player is new to us, then request to be kept informed
+        // of their online status.
+        if (!was_cached) {
+            connect_online(cached.id);
```

Comment:
Mm, this probably isn't a good idea to do for every player we ever see, this can quickly grow to be quite a lot of players when we start thinking about things like viewing tournament results (sometimes hundreds of players), big games on the site (again hundreds of players), every person they've seen in any chat (currently connected or from lingering messages), every custom game that gets created on the play screen pops up a player, all that and coupled with the fact that a lot of people never close their browser, someone could very easily start getting online status updates for hundreds and even thousands of players.

It's this same reason why we have that annoying system in `fetch` to only actually do a call back to the server for more information if we actually need fields that we don't have, most of the interfaces (like the tournament interface, game chat records) will send along enough information about a player for display purposes, but not necessarily enough information for display elsewhere.. so by specifying which fields we need we can prevent having to resolve say 200 different player's on a tournament screen all at once or when a big chat record loads for a game, we have everything we need and can display without hundreds of requests back to the server.